src/BotPlutusInterface/Balance.hs

```diff
@@ -74,62 +78,69 @@ import BotPlutusInterface.Collateral (removeCollateralFromMap)
 import Prettyprinter (pretty, viaShow, (<+>))
 import Prelude
 
--- `BalanceTx` is for checking which type of `Tx` we are balancing.
-data BalanceTx
-  = -- | `BalanceTxCollateral` denotes that we are balancing the `Tx` that creates collateral.
-    BalanceTxCollateral
-  | -- | `BalanceTxWithScripts` denotes that we are balancing the `Tx` that uses scripts.
-    BalanceTxWithScripts
-  | -- | `BalanceTxWithoutScripts` denotes that we are balancing the `Tx` that doesn't uses any scripts.
-    BalanceTxWithoutScripts
-  deriving stock (Eq)
-
--- `balanceTxIO` checks the type of the current `UnbalancedTx` and then calls `balanceTxIO'`.
+data TxWithScript
+data TxWithoutScript
+
+data BalanceTxConstraint a where
+  TxWithScript :: BalanceTxConstraint TxWithScript
+  TxWithoutScript :: BalanceTxConstraint TxWithoutScript
+  TxWithSeparateChange :: BalanceTxConstraint a
+
+instance Eq (BalanceTxConstraint a) where
+  TxWithScript == TxWithScript = True
+  TxWithoutScript == TxWithoutScript = True
+  TxWithSeparateChange == TxWithSeparateChange = True
```

Comment:
this seems a rather involved way of achieving 2 boolean flags, any reason not to just have 2 inputs of binary sum types named accordingly?

Reply:
we can't use normal sum types because, `BalanceTxConstraint a` takes a single type parameter and this is to prevent the following constraints from happening : `balanceTxIO  [TxWithScript, TxWithoutScript]`, currently this constraint will not compile as both `TxWithScript` and `TxWithoutScript` have different types, `BalanceTxConstraint TxWithScript` and `BalanceTxConstraint TxWithoutScript` respectively.

Reply:
Im not saying taking a list at all, just have 2 arguments to the function.
```hs
data ScriptFlag = WithScripts | WithoutScripts
data ChangeFlag = CombineChange | SeparateChange
```
The idea of passing these flags in as a list seems a little odd

Reply:
or even like
```hs
data BalanceConfig = BalanceConfig
  { hasScripts :: Bool
  , separateChange :: Bool
  }
```

Reply:
Yep I think `BalanceConfig` is better. Actually the idea behind list was to implement a way such that multiple constraints can be added while balancing the transaction.

Reply:
We can still acheive that behaviour by extending the record type, which makes for a more traditional and known way of altering behaviour. The list approach seems a little unneccessarily complex

Reply:
I have removed `BalanceTxConstraint` and added `BalanceConfig` as suggested.